File: c_circular_rmq.cpp
#include <bits/stdc++.h>
using namespace std;

struct Segment {
    int n;
    vector<long long> tree, lazy;
    Segment(const vector<int>& nums) {
        n = nums.size();
        tree.assign(4*n, LLONG_MAX);
        lazy.assign(4*n, 0);
        build(0, 0, n-1, nums);
    }
    void build(int pos, int l, int r, const vector<int>& nums) {
        if (l == r) {
            tree[pos] = nums[l];
            return;
        }
        int mid = (l+r)/2;
        build(pos*2+1, l, mid, nums);
        build(pos*2+2, mid+1, r, nums);
        tree[pos] = min(tree[pos*2+1], tree[pos*2+2]);
    }
    void push(int pos, int l, int r) {
        if (lazy[pos] != 0) {
            tree[pos] += lazy[pos];
            if (l != r) {
                lazy[pos*2+1] += lazy[pos];
                lazy[pos*2+2] += lazy[pos];
            }
            lazy[pos] = 0;
        }
    }
    void update(int pos, int l, int r, int L, int R, long long val) {
        push(pos, l, r);
        if (r < L || l > R) return;
        if (l >= L && r <= R) {
            lazy[pos] += val;
            push(pos, l, r);
            return;
        }
        int mid = (l+r)/2;
        update(pos*2+1, l, mid, L, R, val);
        update(pos*2+2, mid+1, r, L, R, val);
        tree[pos] = min(tree[pos*2+1], tree[pos*2+2]);
    }
    long long query(int pos, int l, int r, int L, int R) {
        push(pos, l, r);
        if (r < L || l > R) return LLONG_MAX;
        if (l >= L && r <= R) return tree[pos];
        int mid = (l+r)/2;
        return min(query(pos*2+1, l, mid, L, R), query(pos*2+2, mid+1, r, L, R));
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i=0;i<n;i++) cin >> nums[i];
    int m;
    cin >> m;
    Segment seg(nums);
    while (m--) {
        vector<long long> op;
        string line;
        getline(cin >> ws, line);
        stringstream ss(line);
        long long x;
        while (ss >> x) op.push_back(x);
        int l = op[0], r = op[1];
        if (op.size() == 3) {
            long long val = op[2];
            if (l <= r) seg.update(0, 0, n-1, l, r, val);
            else {
                seg.update(0, 0, n-1, l, n-1, val);
                seg.update(0, 0, n-1, 0, r, val);
            }
        } else {
            if (l <= r) cout << seg.query(0, 0, n-1, l, r) << "\n";
            else cout << min(seg.query(0, 0, n-1, l, n-1), seg.query(0, 0, n-1, 0, r)) << "\n";
        }
    }
}

File: c_circular_rmq.py
class Segment:
    def __init__(self, nums):
        self.nums = nums
        self.n = len(nums)
        self.tree = [float("inf")] * (4 * self.n)
        self.lazy = [0] * (4 * self.n)
        self.build(0, 0, self.n - 1)

    def leftPos(self, pos):
        return (pos << 1) + 1

    def rightPos(self, pos):
        return (pos << 1) + 2

    def build(self, pos, l, r):
        if l == r:
            self.tree[pos] = self.nums[l]
            return

        mid = (l + r) // 2
        self.build(self.leftPos(pos), l, mid)
        self.build(self.rightPos(pos), mid + 1, r)
        self.tree[pos] = min(self.tree[self.leftPos(pos)], self.tree[self.rightPos(pos)])

    def push(self, pos, l, r):
        if self.lazy[pos] != 0:
            self.tree[pos] += self.lazy[pos]
            if l != r:
                self.lazy[self.leftPos(pos)] += self.lazy[pos]
                self.lazy[self.rightPos(pos)] += self.lazy[pos]
            self.lazy[pos] = 0

    def updateHelper(self, pos, l, r, val, L, R):
        self.push(pos, l, r)
        if r < L or l > R:
            return
        if l >= L and r <= R:
            self.lazy[pos] += val
            self.push(pos, l, r)
            return
        mid = (l + r) // 2
        self.updateHelper(self.leftPos(pos), l, mid, val, L, R)
        self.updateHelper(self.rightPos(pos), mid + 1, r, val, L, R)

        self.tree[pos] = min(
            self.tree[self.leftPos(pos)], self.tree[self.rightPos(pos)]
        )

    def queryHelper(self, pos, l, r, L, R):
        self.push(pos, l, r)
        if l > R or r < L:
            return float("inf")
        if l >= L and r <= R:
            return self.tree[pos]
        mid = (l + r) // 2
        return min(
            self.queryHelper(self.leftPos(pos), l, mid, L, R),
            self.queryHelper(self.rightPos(pos), mid + 1, r, L, R),
        )


import sys

# sys.setrecursionlimit(10**6)
from collections import Counter, defaultdict, deque
from bisect import bisect_left, bisect_right
from copy import deepcopy
import math

# input = input
input = sys.stdin.readline


def solve():
    n = int(input())
    nums = list(map(int, input().split()))
    m = int(input())
    seg = Segment(nums)
    for _ in range(m):
        op = list(map(int, input().split()))
        l, r = op[0], op[1]
        if len(op) == 3:
            if l > r:
                seg.updateHelper(0, 0, n - 1, op[2], l, n - 1)
                seg.updateHelper(0, 0, n - 1, op[2], 0, r)
            else:
                seg.updateHelper(0, 0, n - 1, op[2], l, r)
        else:
            if l > r:
                print(
                    min(
                        seg.queryHelper(0, 0, n - 1, l, n - 1),
                        seg.queryHelper(0, 0, n - 1, 0, r),
                    )
                )
            else:
                print(seg.queryHelper(0, 0, n - 1, l, r))


def main():
    t = 1
    for _ in range(t):
        solve()


if __name__ == "__main__":
    main()
